library/src/main/java/com/squareup/timessquare/IndexedLinkedHashMap.java

```diff
@@ -0,0 +1,41 @@
+package com.squareup.timessquare;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Take advantage of LinkedHashMap's iterable ordering but also keep track of the indexes and allow
+ * - Fast key lookup by index
+ * - Fast index lookup by key
+ */
+class IndexedLinkedHashMap<K, V> extends LinkedHashMap<K, V> {
+  private final Map<Integer, K> indexToKey = new LinkedHashMap<>();
+  private final Map<K, Integer> keyToIndex = new LinkedHashMap<>();
```

Comment:
So I just tried this, and unfortunately it breaks on my `getIndexOfKey` method...SparseArray uses object equality, so `indexLookup.indexOfValue(key)` returns -1.  I could iterate over all the keys but that seems worse.  Thoughts?

Diff: https://github.com/square/android-times-square/compare/edenman/faster-highlight...edenman/faster-highlight-with-sparsearray